### PR TITLE
feat: Run cdk-erigon with cdk config files transparently for the user

### PR DESCRIPTION
**Pull request by vcastellm**
## Description

Using the Rust CLI it's now possible to run cdk-erigon that it's present in system's path in a single step providing only the node config file and the genesis file.

```
cdk --config ../kurtosis-cdk/genesis/cdk-node-config.toml --chain ../kurtosis-cdk/genesis/genesis.json erigon
```

Avoiding the necessity of using multiple configuration files and different configuration for `cdk-node` and `cdk-erigon`

image

NOTE: This Rust code is the first iteration into the CLI and can be much improved and refactored in subsequent PRs

---

Steps to test this in local:

1. Requirements: Run your kurtosis environment and have a `cdk-erigon` version installed in your system's PATH
2. `scripts/local_config`
3. `cargo run -- --config ./tmp/cdk/local_config/test.kurtosis.toml --chain ./tmp/cdk/local_config/genesis.json erigon`

### CODE REVIEW

**Review by joanestebanr — COMMENTED** [review]
I don't see how to build the rust client, maybe it needs an entry on Makefile?

**Comment by vcastellm** [comment]
> I don't see how to build the rust client, maybe it needs an entry on Makefile?

It's quite straightforward, in Rust there's the concept of "build scripts" arbitrary commands that can be executed at build time, in this case we use this one https://github.com/0xPolygon/cdk/blob/main/crates/cdk/build.rs, it builds `cdk-node` at build time so both binaries are built and run with just:

```
cargo run
```

 Better entry on README

File: crates/cdk/src/main.rs [hunk 1]
```diff
@@ -6,28 +6,58 @@ use execute::Execute;
...
     let cli = Cli::parse();
 
+    // Read the config
+    let config = read_config(cli.config.clone())?;
+
+    // Initialize the logger
+    logging::tracing(&config.log);
+
+    println!(
+        r#"🐼
```
[the remark this answers is not in the document]
Comment (vcastellm) [reply crates/cdk/src/main.rs:31]: just a nice emoji hahaha, I tend to like emoji logs, can add semantics to the messages, raise awareness and add some cuteness to them. This is just the first one in many? https://www.linkedin.com/pulse/your-application-logs-need-emojis-gianluca-belvisi

File: crates/cdk/src/config_render.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+fn render_yaml(
+    chain_id: String,
+    l2_sequencer_rpc_url: String,
+    datastreamer_host: String,
+    l1_rpc_url: String,
+    sequencer_address: String,
+    zkevm_address: String,
+    rollup_address: String,
+    ger_manager_address: String,
+    matic_contract_address: String,
```
[the remark this answers is not in the document]
Comment (vcastellm) [reply crates/cdk/src/config_render.rs:118]: Not sure, it's indeed the POL token address but it's called `zkevm.l1-matic-contract-address` in erigon config file, let's reduce the debt changing it? or do we wait until we change it everywhere.

File: config/example-config.toml [hunk 3]
```diff
@@ -1,85 +1,101 @@
...
+SequencesTxFileName = "/data/sequencesender.json"
 GasOffset = 80000
 WaitPeriodPurgeTxFile = "15m"
 MaxPendingTx = 1
+MaxBatchesForL1 = 300
+BlockFinality="SafeBlock" # TODO: should be finalized but it breaks :)
+SanityCheckRPCURL = ""
 	[SequenceSender.StreamClient]
-		Server = "127.0.0.1:6900"
+		Server = "cdk-erigon-sequencer-001:6900"
```
[the remark this answers is not in the document]
Comment (vcastellm) [reply config/example-config.toml]: Removed the file

File: crates/cdk-config/src/sequence_sender.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+use serde::Deserialize;
+
+/// The SequenceSender configuration.
+#[derive(Deserialize, Debug, Clone)]
+pub struct SequenceSender {
```
Comment (Stefan-Ethernal) [inline crates/cdk-config/src/sequence_sender.rs:5]: Are we missing the `EthTxManager` configuration here?

File: crates/cdk-config/src/telemetry.rs [hunk 5]
```diff
@@ -1,14 +1,13 @@
...
+use serde::Deserialize;
+use std::net::SocketAddr;
 
 #[derive(Deserialize, Debug, Clone, Copy)]
 #[serde(rename_all = "PascalCase")]
+#[allow(dead_code)]
 pub struct TelemetryConfig {
     #[serde(rename = "PrometheusAddr", default = "default_metrics_api_addr")]
-    pub addr: SocketAddr,
+    pub addr: SocketAddr, //
```
Comment (Stefan-Ethernal) [inline crates/cdk-config/src/telemetry.rs:10]: Missing comment? Should we remove the slashes?
```suggestion
    pub addr: SocketAddr,
```

**Review by Stefan-Ethernal — COMMENTED** [review]
Generally speaking, should we match the current configuration of the CDK, as we are most likely missing some fields (`EthTxManager` for example)?

**Comment by vcastellm** [comment]
> Generally speaking, should we match the current configuration of the CDK, as we are most likely missing some fields (`EthTxManager` for example)?

Good observation @Stefan-Ethernal in general we're not adding all config values just some at this point, some of them we use it to construct the config, others are simply missing and others are there to use them for status checking in the future. I preferred just leaving them as is instead of removing them, we can refactor them in upcoming PRs in case the usage we're going to give them.

File: crates/cdk-config/src/aggregator.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+#[derive(Deserialize, Debug, Clone)]
+pub struct StreamClient {
+    #[serde(rename = "Server")]
+    pub server: String,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct EthTxManager {
+    #[serde(rename = "Etherman")]
+    pub etherman: Etherman,
```
[the remark this answers is not in the document]
Comment (vcastellm) [reply crates/cdk-config/src/aggregator.rs:15]: See comment

**Review by Stefan-Ethernal — APPROVED** [review]

**Review by rbpol — APPROVED** [review]

**Pull request merged**